File: androidApp/src/main/java/com/smp/moviediary/androidApp/mapper/MovieMapper.kt
package com.smp.moviediary.androidApp.mapper

import androidx.core.text.HtmlCompat
import com.smp.moviediary.androidApp.model.Movie
import com.smp.moviediary.shared.domain.entity.MovieEntity
import com.smp.moviediary.shared.domain.mapper.Mapper


object MovieMapper : Mapper<MovieEntity, Movie> {

    override fun map(data: MovieEntity): Movie {
        return Movie(
            id = data.id,
            title = data.title,
            spannedTitle = HtmlCompat.fromHtml(data.title, HtmlCompat.FROM_HTML_MODE_LEGACY),
            subTitle = data.subTitle,
            spannedSubTitle = data.subTitle?.let { subTitle ->
                HtmlCompat.fromHtml(subTitle, HtmlCompat.FROM_HTML_MODE_LEGACY)
            },
            thumb = data.thumb,
            director = data.director,
            actor = data.actor,
            publishedAt = data.publishedAt,
            rating = data.rating,
            scrap = data.scrap,
            memo = data.memo
        )
    }

    override fun reverse(data: Movie): MovieEntity {
        return MovieEntity(
            id = data.id,
            title = data.title,
            subTitle = data.subTitle,
            thumb = data.thumb,
            director = data.director,
            actor = data.actor,
            publishedAt = data.publishedAt,
            rating = data.rating,
            scrap = data.scrap,
            memo = data.memo
        )
    }
}

File: iosApp/iosApp/MovieUI/Model/MovieMapper.swift
//
//  MovieMapper.swift
//  iosApp
//
//  Created by myungpyo on 2021/01/10.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import shared

class MovieMapper: Mapper {
    typealias From = MovieEntity
    typealias To = Movie
    
    static let shared = MovieMapper()
    
    func map(data: MovieEntity) -> Movie {
        Movie(
            id: data.id,
            title: data.title,
            attributedTitle: makeHtmlString(text: data.title),
            subTitle: data.subTitle,
            attributedSubtitle: makeHtmlStringOrNil(text: data.subTitle),
            thumb: data.thumb,
            director: data.director,
            actor: data.actor,
            publishedAt: data.publishedAt,
            rating: data.rating,
            scrap: data.scrap,
            memo: data.memo
        )
    }
    
    func reverse(data: Movie) -> MovieEntity {
        MovieEntity(
            id: data.id,
            title: data.title,
            subTitle: data.subTitle,
            thumb: data.thumb,
            director: data.director,
            actor: data.actor,
            publishedAt: data.publishedAt,
            rating: data.rating,
            scrap: data.scrap,
            memo: data.memo
        )
    }
    
    private func makeHtmlString(text: String) -> NSAttributedString {
        return makeHtmlStringOrNil(text: text)!
    }
    
    private func makeHtmlStringOrNil(text: String?) -> NSAttributedString? {
        guard let data = text?.data(using: String.Encoding.unicode) else { return nil }
        return try? NSAttributedString(data: data,
                           options: [.documentType:NSAttributedString.DocumentType.html],
                documentAttributes: nil)
    }
}
